### PR TITLE
[addons] change ADDON_GLOBAL_VERSION_MAIN to 1.1.0 and leave min by old 1.0.12

### PR DESCRIPTION
**Pull request by AlwinEsch**
##  Description
This to bring backward compatibility. Related to https://github.com/xbmc/xbmc/pull/16401 and https://github.com/xbmc/xbmc/pull/16444

## Types of change

- [ ] **Bug fix** (non-breaking change which fixes an issue)
- [ ] **Clean up** (non-breaking change which removes non-working, unmaintained functionality)
- [ ] **Improvement** (non-breaking change which improves existing functionality)
- [ ] **New feature** (non-breaking change which adds functionality)
- [ ] **Breaking change** (fix or feature that will cause existing functionality to change)
- [ ] **Cosmetic change** (non-breaking change that doesn't touch code)
- [ ] **None of the above** (please explain below)

## Checklist:

- [ ] My code follows the **[Code Guidelines](https://github.com/xbmc/xbmc/blob/master/docs/CODE_GUIDELINES.md)** of this project 
- [ ] My change requires a change to the documentation, either Doxygen or wiki
- [ ] I have updated the documentation accordingly
- [ ] I have read the **[Contributing](https://github.com/xbmc/xbmc/blob/master/docs/CONTRIBUTING.md)** document
- [ ] I have added tests to cover my change
- [ ] All new and existing tests passed

### CODE REVIEW

**Comment by ksooo** [comment]
What about the changes in "xbmc/addons/interfaces/General.h" you did with #16401? Shouldn't the two new functions moved to the end of the function table as well?

File: xbmc/addons/interfaces/General.cpp [hunk 1]
```diff
@@ -55,10 +55,10 @@ void Interface_General::Init(AddonGlobalInterface* addonInterface)
   addonInterface->toKodi->kodi->get_region = get_region;
   addonInterface->toKodi->kodi->get_free_mem = get_free_mem;
   addonInterface->toKodi->kodi->get_global_idle_time = get_global_idle_time;
+  addonInterface->toKodi->kodi->kodi_version = kodi_version;
```
Comment (Rechi) [inline xbmc/addons/interfaces/General.cpp]: Is moving this line necessary? If it is, this is the wrong place.

File: xbmc/addons/interfaces/General.cpp [hunk 2]
```diff
@@ -409,6 +409,47 @@ int Interface_General::get_global_idle_time(void* kodiBase)
   return g_application.GlobalIdleTime();
 }
 
+void Interface_General::kodi_version(void* kodiBase, char** compile_name, int* major, int* minor, char** revision, char** tag, char** tagversion)
```
Comment (Rechi) [inline xbmc/addons/interfaces/General.cpp]: Is moving this function necessary? If it is, this is the wrong place.

File: xbmc/addons/interfaces/General.h [hunk 3]
```diff
@@ -52,10 +52,10 @@ namespace ADDON
     static char* get_region(void* kodiBase, const char* id);
     static void get_free_mem(void* kodiInstance, long* free, long* total, bool as_bytes);
     static int  get_global_idle_time(void* kodiBase);
+    static void kodi_version(void* kodiBase, char** compile_name, int* major, int* minor, char** revision, char** tag, char** tagversion);
```
Comment (Rechi) [inline xbmc/addons/interfaces/General.h]: This is the wrong place.

**Review by Rechi — COMMENTED** [review]
Add-ons which where compiled after #16401, but before this PR will now show as compatible, but are actually broken.

File: xbmc/addons/interfaces/General.cpp (hunk 1 above)
Comment (AlwinEsch) [reply xbmc/addons/interfaces/General.cpp]: Not direct necessary but to have for better overview in same flow as https://github.com/xbmc/xbmc/blob/master/xbmc/addons/kodi-addon-dev-kit/include/kodi/General.h#L47-L64

File: xbmc/addons/interfaces/General.cpp (hunk 2 above)
Comment (AlwinEsch) [reply xbmc/addons/interfaces/General.cpp]: Same as https://github.com/xbmc/xbmc/pull/16455/files#r311907505

File: xbmc/addons/interfaces/General.h (hunk 3 above)
Comment (AlwinEsch) [reply xbmc/addons/interfaces/General.h]: This to bring backward compatibility

Comment (Rechi) [reply xbmc/addons/interfaces/General.h]: I know, but `kodi_version` was after `get_current_skin_id` in this struct.

**Comment by AlwinEsch** [comment]
Where the funtions in https://github.com/xbmc/xbmc/blob/master/xbmc/addons/interfaces/General.h and https://github.com/xbmc/xbmc/blob/master/xbmc/addons/interfaces/General.cpp are secondary, important is the `external 'C"` to have functions in the right format.
The address of the individual is always loaded at addon start and with table https://github.com/xbmc/xbmc/blob/master/xbmc/addons/kodi-addon-dev-kit/include/kodi/General.h#L47-L64 passed.

File: xbmc/addons/interfaces/General.h (hunk 3 above)
Comment (AlwinEsch) [reply xbmc/addons/interfaces/General.h]: No was before https://github.com/xbmc/xbmc/blob/4398f988e1577495ab6a9ba1e43e6bd2ac65f465/xbmc/addons/kodi-addon-dev-kit/include/kodi/General.h#L37-L38 the `get_current_skin_id` was in newer versions added, thats why behind.

Comment (Rechi) [reply xbmc/addons/interfaces/General.h]: You looked at / linked to the wrong file, this is `xbmc/addons/interfaces/General.h` not `xbmc/addons/kodi-addon-dev-kit/include/kodi/General.h`.

**Review by ksooo — CHANGES_REQUESTED** [review]
Let's please revert function order to state before #16401

TBH, I see absolutely no benefit in this (despite from cosmetical POV) in resorting. OTOH it breaks ABI for no real reason.

I suggest to put those cleanup-only changes on a list and to do those when an incompatible API change is unavoidable.

File: xbmc/addons/interfaces/General.h (hunk 3 above)
Comment (AlwinEsch) [reply xbmc/addons/interfaces/General.h]: True, mix these more often in my head. But think of sorting it right there in a dishwashing, not bad.

File: xbmc/addons/interfaces/General.cpp (hunk 1 above)
Comment (ksooo) [reply xbmc/addons/interfaces/General.cpp]: > Not direct necessary

 Can we then please change this later, with the next change where ABI breakage is really unavoidable?

Comment (ksooo) [reply xbmc/addons/interfaces/General.cpp]: order has to be then:
* get_global_idle_time
* get_current_skin_id
* kodi_version
* change_keyboard_layout
* get_keyboard_layout

Comment (ksooo) [reply xbmc/addons/interfaces/General.cpp]: deleted my previous comments... was the wrong place.

File: xbmc/addons/interfaces/General.h (hunk 3 above)
Comment (ksooo) [reply xbmc/addons/interfaces/General.h]: > But think of sorting it right there in a dishwashing, not bad

 Yes, but can we please make this change later, in another PR?

**Review by ksooo — APPROVED** [review]
After discussing with @AlwinEsch internally, we came to the conclusion that the changes (especially resorting functions for new addon API) are okay, because the underlying C-API that was incompatibly changed in this PR is currently not public. Thanks for explaining things, Alwin.

**Comment by ksooo** [comment]
jenkins build this please

**Comment by ksooo** [comment]
> Add-ons which where compiled after #16401, but before this PR will now show as compatible, but are actually broken.

 Maybe this is something for a Followup-PR, @AlwinEsch

**Comment by AlwinEsch** [comment]
> > Add-ons which where compiled after #16401, but before this PR will now show as compatible, but are actually broken.
> 
> Maybe this is something for a Followup-PR, @AlwinEsch

In principle, correct. However, the changed parts are still not used anywhere (so in no repo addons).
Therefore it will not be noticeable on addons.

**Comment by Rechi** [comment]
This is nothing which can be fixed in a PR without adjusting min version. Also it's not about the not yet used functions, it's about the different offset of the functions which existed before.

**Comment by ksooo** [comment]
You are right @Rechi , but do we actually have "Add-ons which where compiled after #16401, but before this PR" in the wild? To prevent this I was in a real hurry to merge this PR.

**Comment by Rechi** [comment]
Who is `we` and why not care about everyone?

**Comment by ksooo** [comment]
???

**Comment by ksooo** [comment]
> This is nothing which can be fixed in a PR without adjusting min version. Also it's not about the not yet used functions, it's about the different offset of the functions which existed before.

IMO this is about a lack of functionality in Kodi's comaptibility checking code. Kodi must refuse to load an addon with API versions greater than own versions.

**Comment by Rechi** [comment]
`1.0.15` < `1.1.0`, so Kodi would behave correct.

**Comment by ksooo** [comment]
> 1.0.15 < 1.1.0, so Kodi would behave correct.

 Ah, now I get it. "Add-ons which where compiled after #16401, but before this PR" require 1.0.15 and Kodi compiled with this PR merged has version 1.1.0 and min version 1.0.12 - So the addon is considered compatible. Damn.

> Who is we and why not care about everyone?

 I consider this a one-time alpha state faux pas whcih never ever must repeat.

**Pull request merged**